src/Http/Routing/src/Matching/ContentEncodingNegotiationMatcherPolicy.cs

```diff
@@ -0,0 +1,123 @@
+
+    internal class ContentEncodingPolicyJumpTable(int anyContentEncodingDestination, int noContentEncodingDestination, ContentEncodingDestinationsLookUp destinations) : NegotiationPolicyJumpTable("Accept-Encoding", anyContentEncodingDestination, noContentEncodingDestination)
+    {
+        private readonly ContentEncodingDestinationsLookUp _destinations = destinations;
+
+        protected override int GetDestination(string? value) => _destinations.GetDestination(value);
+
+        protected override double GetQuality(string? value) => _destinations.GetValueQuality(value);
+    }
+
+    internal class ContentEncodingDestinationsLookUp
+    {
+        private readonly int _brotliDestination = -1;
+        private readonly double _brotliQuality;
+        private readonly int _gzipDestination = -1;
+        private readonly double _gzipQuality;
+        private readonly int _identityDestination = -1;
+        private readonly double _identityQuality;
+        private readonly Dictionary<string, (int destination, double quality)>? _extraDestinations;
+
+        public ContentEncodingDestinationsLookUp((string contentEncoding, double quality, int destination)[] destinations)
+        {
+            for (var i = 0; i < destinations.Length; i++)
+            {
+                var (contentEncoding, quality, destination) = destinations[i];
+                switch (contentEncoding.ToLowerInvariant())
+                {
+                    case "br":
+                        _brotliDestination = destination;
+                        _brotliQuality = quality;
+                        break;
+                    case "gzip":
+                        _gzipDestination = destination;
+                        _gzipQuality = quality;
+                        break;
+                    case "identity":
+                        _identityDestination = destination;
+                        _identityQuality = quality;
+                        break;
+                    default:
+                        _extraDestinations ??= new Dictionary<string, (int destination, double quality)>(StringComparer.OrdinalIgnoreCase);
+                        _extraDestinations.Add(contentEncoding, (destination, quality));
+                        break;
+                }
+            }
+        }
+
+        public int GetDestination(string? negotiationValue)
+        {
+            var (matchedEncoding, destination) = negotiationValue?.Length switch
+            {
+                2 => ("br", _brotliDestination),
+                4 => ("gzip", _gzipDestination),
+                8 => ("identity", _identityDestination),
+                _ => (null, -1)
+            };
```

Comment:
I'm guessing that this was done for perf reasons? If so, I wonder if we could have a comment here explaining that the `matchedEncoding` will get validated later in the method. I initially thought there was a potential bug here because the `zstd` content encoding also has 4 characters, but then read further and realized that we later verify that `negotiationValue` and `matchedEncoding` are equivalent.

Reply:
@MackinnonBuck yes, consider these tables as a very special cased way of comparing values coming from the request to values, we know exist to accelerate matching. See https://github.com/dotnet/aspnetcore/blob/0e7349e41a112a0c62f364ce9666a1a7dd4dc2fd/src/Http/Routing/src/Matching/HttpMethodDestinationsLookup.cs for another example.